index.js

```diff
@@ -20,302 +36,589 @@ var prefs = Cc["@mozilla.org/preferences-service;1"].getService(Ci.nsIPrefServic
+            dlog("Error loading ticker " + tickerId + 
+                          ". URL is not correctly responding:" + url);
+            return;
           }
-          var price = response.json;
-          for (var i = 0; i < jsonPath.length; i++) { // Parse JSON path
-            if (typeof price[jsonPath[i]] == "undefined") {
-              if (DEBUG) {
-                console.log(TAG + " error loading ticker " + id + 
-                            ". URL is not correctly responding:" + url);
+          price = price[jsonPath[i]];
+        }
+        dlog("Price received and parsed for " + tickerId + ": " + price);
+        var doc = tickers[tickerId].doc;
+        if (usingWidgets && 
+            (doc !== undefined) && (doc !== null)) { // For Widgets
+          var win = getWidgetWindow(tickerId);
+          if (win) {
+            win.postMessage({
+              "type": "updateTickerModelPrice",
+              "id": tickerId,
+              "data": {
+                "price": price
               }
-              return;
-            }
-            price = price[jsonPath[i]];
-          }
-          if (DEBUG) {
-            console.log(TAG + " Price received and parsed: " + price);
+            }, "*");
+            adjustWidgetSize(tickerId);
           }
+        } else if (tickersFrame !== null) { // For Toolbar
           tickersFrame.postMessage({
             "type": "updateTickerModelPrice",
-            "id": id,
+            "id": tickerId,
             "data": {
               "price": price
             }
           }, tickersFrame.url);
+        } else {
+          dlog("Frame and Widget document are both empty for Widget " + tickerId);
         }
       }
-    }).get();
-  }
+    }
+  }).get();
+}
+
+
+
+// Methods to create tickers and manipulate them
+
+
+/**
+ * Create a Widget, attach listeners and trigger update on Widget Added to window
+ **/
+function createNewTickersWidget(tickerId) {
+  dlog("Creating widget for ticker " + tickerId);
+  var tickerTitle = tickers[tickerId].exchangeName + " " + 
+                      tickers[tickerId].currency + "/" + tickers[tickerId].baseCurrency
+  CustomizableUI.createWidget({
+    id:           tickerId + WIDGET_SUFFIX,
+    type:         "custom",
+    label:        tickerId,
+    removable:    true,
+    defaultArea:  CustomizableUI.AREA_NAVBAR,
+    /**
+     * Build widget iFrame content and attach listeners
+     **/
+    onBuild: function(aDocument) {
+      tickers[tickerId].doc = aDocument;
+      var node = aDocument.createElement("toolbaritem");
+      var props = {
+        id:           this.id,
+        title:        "Bitcoin Price Ticker " + tickerId,
+        align:        "center",
+        label:        tickerTitle,
+        width:        "15px",
+        height:       10,
+        tooltiptext:  tickerTitle,
+        class:        "chromeclass-toolbar-additional panel-wide-item"
+      };
+      for (var p in props) {
```

Comment:
The body of a for in should be wrapped in an if statement to filter unwanted properties from the prototype.